In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler 
from sklearn.preprocessing import OneHotEncoder

import statsmodels.api as sm

# Linear model
from sklearn import linear_model                     # Linear model
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error, r2_score


# Logistic Regression model
from sklearn.linear_model import LogisticRegression  # Logistic Regression model
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score


import sklearn
from sklearn.utils import resample

### Instructions

At this point, we have created a model to predict who will make a donation and who won't (Classification Model). But, what about the ammount of money that each person will give?

In this lab, subset those that have made a donation (Target B) and use that subset to create a model to predict how much money will they give (Target D) (Regression Model).

- Only look at people who have donated (Target B = 1)
- Use this new dataframe to create a model to predict how much they will donate (Target D)
- Using the regression model, make predictions on all of the people our classification model predicted will donate.
- See the pdf file for a schema of the process.
Evaluate the result of your model and estimate how much better the result are for the business in comparison with the naive scenario we discuss on Monday. (Just sending donation cards to everyone)

#### Loading original data

In [2]:
# use original dataset
numerical = pd.read_csv('numerical.csv')
targets = pd.read_csv('target.csv')
categorical = pd.read_csv('categorical.csv')

In [3]:
all_data = pd.concat((numerical, categorical,targets),axis=1)

In [4]:
all_data.shape

(95412, 339)

#### Subset only Target_B = 1

In [5]:
donations_data = all_data[all_data['TARGET_B']==1]
X = donations_data.drop(columns=['TARGET_B','TARGET_D'])
y = donations_data['TARGET_D']

In [6]:
display(X.shape)
y.shape

(4843, 337)

(4843,)

In [7]:
pd.set_option('display.max_columns',None)
X.head()

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
20      2  62.000000       3        8   10         2       25        40   
30      0  61.611649       5        9    0         1       37        58   
45      0  66.000000       5        9    5         0       33        24   
78      0  69.000000       6        9    0         0       34        20   
93      1  73.000000       1        7   10         0       21        53   

    WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
20        27        11         4       1        9    2707     672     929   
30        16         8         1       5        9    2147     591     640   
45        39         6         5       1        9    2160     683     900   
78        54         2         3       1        9   13801    3736    6388   
93         8         5         4      11        7    1673     418     462   

    POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  ETH5  \
20       99        0        0       45       55    97     1     0     2     1   
30       99        0        0       49       51    94     2     0     3     5   
45       89        0       11       48       52    99     0     0     1     1   
78       99        0        0       48       52    97     0     0     3     3   
93       99        0        0       49       51     7    93     0     0     0   

    ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  ETH16  \
20     0     0     0     0      1      0      0      1      0      0      0   
30     1     0     1     1      0      0      0      3      0      0      2   
45     0     0     0     0      0      0      0      0      0      0      0   
78     0     1     1     0      0      0      0      2      0      0      1   
93     0     0     0     0      0      0      0      0      0      0      0   

    AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  CHIL2  \
20      35      42      43      37      47      50      27     34     43   
30      29      36      37      27      38      40      37     46     37   
45      42      47      49      40      48      50      20     33     42   
78      40      44      46      42      47      49      14     40     37   
93      29      39      42      31      40      45      30     32     41   

    CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  CHILC2  \
20     23      9     20     27     16      8      6     15      15      13   
30     17      8     34     34     15      6      2      1      18      21   
45     25      7     14     23     21     17     13      5      14      13   
78     23      8     21     21     17     13     13      7      18      15   
93     26     18     18     29     20      9      4      3      14      13   

    CHILC3  CHILC4  CHILC5  HHAGE1  HHAGE2  HHAGE3  HHN1  HHN2  HHN3  HHN4  \
20      37      21      14      16       8      15    23    30    47    29   
30      31      19      11       5       0       3     7    23    71    47   
45      33      22      18      24       7      22    22    43    35    19   
78      30      19      18      26       9      24    29    44    27    13   
93      29      26      18      14       2      10     8    21    72    50   

    HHN5  HHN6  MARR1  MARR2  MARR3  MARR4  HHP1  HHP2  DW1  DW2  DW3  DW4  \
20    11     2     55     10     13     22   190   267   89   49    1   10   
30    15     6     77      5      1     17   286   334   99   99    0    0   
45     6     1     72      7      5     16   165   240   69   58    1   28   
78     4     1     54     13      6     27   148   216   78   58    8   20   
93    24    13     47     15      4     34   300   358   98   98    0    0   

    DW5  DW6  DW7  DW8  DW9   HV1   HV2  HV3  HV4  HU1  HU2  HU3  HU4  HU5  \
20   10    8    8    8    0   902   960    8    8   88   12   97    3   11   
30    0    0    0    0    0  1002  1166    7    5   97    3   96    4    3   
45   27    4    0    0    0  1

In [8]:
#### 

In [9]:
X.head()

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
20      2  62.000000       3        8   10         2       25        40   
30      0  61.611649       5        9    0         1       37        58   
45      0  66.000000       5        9    5         0       33        24   
78      0  69.000000       6        9    0         0       34        20   
93      1  73.000000       1        7   10         0       21        53   

    WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
20        27        11         4       1        9    2707     672     929   
30        16         8         1       5        9    2147     591     640   
45        39         6         5       1        9    2160     683     900   
78        54         2         3       1        9   13801    3736    6388   
93         8         5         4      11        7    1673     418     462   

    POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  ETH5  \
20       99        0        0       45       55    97     1     0     2     1   
30       99        0        0       49       51    94     2     0     3     5   
45       89        0       11       48       52    99     0     0     1     1   
78       99        0        0       48       52    97     0     0     3     3   
93       99        0        0       49       51     7    93     0     0     0   

    ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  ETH16  \
20     0     0     0     0      1      0      0      1      0      0      0   
30     1     0     1     1      0      0      0      3      0      0      2   
45     0     0     0     0      0      0      0      0      0      0      0   
78     0     1     1     0      0      0      0      2      0      0      1   
93     0     0     0     0      0      0      0      0      0      0      0   

    AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  CHIL2  \
20      35      42      43      37      47      50      27     34     43   
30      29      36      37      27      38      40      37     46     37   
45      42      47      49      40      48      50      20     33     42   
78      40      44      46      42      47      49      14     40     37   
93      29      39      42      31      40      45      30     32     41   

    CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  CHILC2  \
20     23      9     20     27     16      8      6     15      15      13   
30     17      8     34     34     15      6      2      1      18      21   
45     25      7     14     23     21     17     13      5      14      13   
78     23      8     21     21     17     13     13      7      18      15   
93     26     18     18     29     20      9      4      3      14      13   

    CHILC3  CHILC4  CHILC5  HHAGE1  HHAGE2  HHAGE3  HHN1  HHN2  HHN3  HHN4  \
20      37      21      14      16       8      15    23    30    47    29   
30      31      19      11       5       0       3     7    23    71    47   
45      33      22      18      24       7      22    22    43    35    19   
78      30      19      18      26       9      24    29    44    27    13   
93      29      26      18      14       2      10     8    21    72    50   

    HHN5  HHN6  MARR1  MARR2  MARR3  MARR4  HHP1  HHP2  DW1  DW2  DW3  DW4  \
20    11     2     55     10     13     22   190   267   89   49    1   10   
30    15     6     77      5      1     17   286   334   99   99    0    0   
45     6     1     72      7      5     16   165   240   69   58    1   28   
78     4     1     54     13      6     27   148   216   78   58    8   20   
93    24    13     47     15      4     34   300   358   98   98    0    0   

    DW5  DW6  DW7  DW8  DW9   HV1   HV2  HV3  HV4  HU1  HU2  HU3  HU4  HU5  \
20   10    8    8    8    0   902   960    8    8   88   12   97    3   11   
30    0    0    0    0    0  1002  1166    7    5   97    3   96    4    3   
45   27    4    0    0    0  1

In [10]:
y.head()

20     4.0
30     7.0
45     5.0
78    13.0
93    10.0
Name: TARGET_D, dtype: float64

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

numericalX_train   = X_train.select_dtypes(np.number)
numericalX_test    = X_test.select_dtypes(np.number)
categoricalX_train = X_train.select_dtypes(object)
categoricalX_test  = X_test.select_dtypes(object)

# we OneHotEncode the categoricals 
encoder = OneHotEncoder(drop='first').fit(categoricalX_train)
encoded_categorical_train = encoder.transform(categoricalX_train).toarray()
encoded_categorical_train = pd.DataFrame(encoded_categorical_train, columns=encoder.get_feature_names_out()) # needed to avoid error
encoded_categorical_test = encoder.transform(categoricalX_test).toarray()
encoded_categorical_test = pd.DataFrame(encoded_categorical_test, columns=encoder.get_feature_names_out())

# Scalling numerical
transformer = MinMaxScaler().fit(numericalX_train)
scaled_numerical_train = transformer.transform(numericalX_train)
scaled_numerical_train = pd.DataFrame(scaled_numerical_train, columns=numericalX_train.columns)
scaled_numerical_test  = transformer.transform(numericalX_test)
scaled_numerical_test  = pd.DataFrame(scaled_numerical_test, columns=numericalX_train.columns)

X_train_treated = pd.concat([scaled_numerical_train, encoded_categorical_train], axis = 1)
X_test_treated  = pd.concat([scaled_numerical_test,  encoded_categorical_test],  axis = 1)

In [12]:
# X_train_treated.isna().sum()
# no null values

In [13]:
X_train_treated.shape

(3874, 354)

#### Creating linear regression model on all features

- first I will performe a model on all features, as my experience from building the last model was, that reducing the features didnt improve the model, but even made it performe slightly worst.


In [14]:
# Creating Linear Regression Model

lm = linear_model.LinearRegression()
lm.fit(X_train_treated,y_train)

LinearRegression()

In [15]:
predictions_train = lm.predict(X_train_treated) # y_train_pred
predictions_test = lm.predict(X_test_treated)   # y_test_pred

In [16]:
# For Linear Regression (linear model)

r2_score_train = r2_score(y_train, predictions_train)
r2_score_test = r2_score(y_test, predictions_test)

print("R2_Score for train: ", r2_score_train)  #r2_score shows how well my model can predict 
print("R2_Score for test: ", r2_score_test)

R2_Score for train:  0.6095586890061206
R2_Score for test:  0.2895529501343084


In [17]:
# R2-Score shows the accuracy of the model. 
# The model is working ok on a train set, but bad on test set. 
# Overall the model is not performing very well on a test data set. 

In [18]:
from sklearn.metrics import mean_squared_error as mse

mse_train = mean_squared_error(y_train, predictions_train)
mse_test = mean_squared_error(y_test, predictions_test)

print("Mean Squared Error for train: ", mse_train)
print("Mean Squared Error for test: ", mse_test)

Mean Squared Error for train:  56.834907740715956
Mean Squared Error for test:  135.38611017671997


In [19]:
# The MSE for the training set is lower than the MSE for the test set and the difference is very big. 
# Overall the model is not performing that well. 

### Predictions
- here I make predictions for the amount of the donated money.
- This will be performed only on the donors, who were predicted with the classification model from the previous lab. 


In [20]:
# Importing predicted TARGET_B_predicted
donors_predicted = pd.read_csv('donors_predicted.csv')

In [21]:
donors_predicted

TARGET_B_predicted
0                       1
1                       0
2                       1
3                       1
4                       0
...                   ...
95407                   0
95408                   0
95409                   0
95410                   1
95411                   0

[95412 rows x 1 columns]

In [22]:
# Creating a dataset to apply this on


all_data.head()
X = all_data.drop(["TARGET_B", "TARGET_D"], axis=1)
X.head()

TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  \
0      0  60.000000       5        9    0         0       39        34   
1      1  46.000000       6        9   16         0       15        55   
2      1  61.611649       3        1    2         0       20        29   
3      0  70.000000       1        4    2         0       23        14   
4      0  78.000000       3        2   60         1       28         9   

   WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  \
0        18        10         2       1        5     992     264     332   
1        11         6         2       1        9    3611     940     998   
2        33         6         8       1        1    7001    2040    2669   
3        31         3         0       3        0     640     160     219   
4        53        26         3       2        9    2520     627     761   

   POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  ETH5  \
0        0       35       65       47       53    92     1     0     0    11   
1       99        0        0       50       50    67     0     0    31     6   
2        0        2       98       49       51    96     2     0     0     2   
3        0        8       92       54       46    61     0     0    11    32   
4       99        0        0       46       54     2    98     0     0     1   

   ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  ETH16  \
0     0     0     0     0      0      0      0     11      0      0      0   
1     4     2     6     4     14      0      0      2      0      1      4   
2     0     0     0     0      0      0      0      2      0      0      0   
3     6     2     0     0      0      0      0     31      0      0      1   
4     0     0     0     0      0      0      0      0      0      0      0   

   AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  CHIL2  \
0      39      48      51      40      50      54      25     31     42   
1      34      41      43      32      42      45      32     33     46   
2      35      43      46      37      45      49      23     35     40   
3      32      40      44      34      43      47      25     45     35   
4      33      45      50      36      46      50      27     34     43   

   CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  CHILC2  \
0     27     11     14     18     17     13     11     15      12      11   
1     21     13     14     33     23     10      4      2      11      16   
2     25     13     20     19     16     13     10      8      15      14   
3     20     15     25     17     17     12      7      7      20      17   
4     23     14     21     13     15     20     12      5      13      15   

   CHILC3  CHILC4  CHILC5  HHAGE1  HHAGE2  HHAGE3  HHN1  HHN2  HHN3  HHN4  \
0      34      25      18      26      10      23    18    33    49    28   
1      36      22      15      12       1       5     4    21    75    55   
2      30      22      19      25      10      23    21    35    44    22   
3      30      14      19      25      11      23    23    27    50    30   
4      34      19      19      31       7      27    16    26    57    36   

   HHN5  HHN6  MARR1  MARR2  MARR3  MARR4  HHP1  HHP2  DW1  DW2  DW3  DW4  \
0    12     4     61      7     12     19   198   276   97   95    2    2   
1    23     9     69      4      3     24   317   360   99   99    0    0   
2     6     2     63      9      9     19   183   254   69   69    1    6   
3    15     8     63      9      6     23   199   283   85   83    3    4   
4    24    14     42     17      9     33   235   323   99   98    0    0   

   DW5  DW6  DW7  DW8  DW9   HV1   HV2  HV3  HV4  HU1  HU2  HU3  HU4  HU5  \
0    0    0    7    7    0   479   635    3    2   86   14   96    4    7   
1    0    0    0    0    0  5468  5218   12   10   96    4   97    3    9   
2    5    3    3    3    0   497   546    2    1   78   22   93    7   18   
3  

In [23]:
numericalX    = X.select_dtypes(np.number)
categoricalX = X.select_dtypes(object)

encoded_categorical_X = encoder.transform(categoricalX).toarray()
encoded_categorical_X_df=pd.DataFrame(encoded_categorical_X, columns=encoder.get_feature_names_out())

scaled_numerical_X = transformer.transform(numericalX)
scaled_numerical_X_df= pd.DataFrame(scaled_numerical_X, columns=numericalX.columns)

X_df = pd.concat([scaled_numerical_X_df, encoded_categorical_X_df], axis = 1)

In [24]:
X_df

TCODE       AGE    INCOME   WEALTH1       HIT  MALEMILI   MALEVET  \
0      0.000000  0.542169  0.666667  1.000000  0.000000  0.000000  0.493671   
1      0.000026  0.373494  0.833333  1.000000  0.066667  0.000000  0.189873   
2      0.000026  0.561586  0.333333  0.111111  0.008333  0.000000  0.253165   
3      0.000000  0.662651  0.000000  0.444444  0.008333  0.000000  0.291139   
4      0.000000  0.759036  0.333333  0.222222  0.250000  0.010417  0.354430   
...         ...       ...       ...       ...       ...       ...       ...   
95407  0.000026  0.561586  0.666667  1.000000  0.000000  0.145833  0.455696   
95408  0.000026  0.397590  1.000000  1.000000  0.004167  0.000000  0.392405   
95409  0.000026  0.542169  0.666667  1.000000  0.000000  0.000000  0.227848   
95410  0.000000  0.518072  1.000000  1.000000  0.000000  0.000000  0.354430   
95411  0.000051  0.783133  0.666667  0.888889  0.012500  0.010417  0.417722   

       VIETVETS  WWIIVETS  LOCALGOV  STATEGOV    FEDGOV   WEALTH2    POP901  \
0      0.343434  0.181818  0.208333      0.04  0.016667  0.555556  0.016206   
1      0.555556  0.111111  0.125000      0.04  0.016667  1.000000  0.058993   
2      0.292929  0.333333  0.125000      0.16  0.016667  0.111111  0.114375   
3      0.141414  0.313131  0.062500      0.00  0.050000  0.000000  0.010456   
4      0.090909  0.535354  0.541667      0.06  0.033333  1.000000  0.041169   
...         ...       ...       ...       ...       ...       ...       ...   
95407  0.474747  0.111111  0.145833      0.16  0.216667  1.000000  0.447305   
95408  0.434343  0.191919  0.083333      0.02  0.000000  1.000000  0.020487   
95409  0.464646  0.202020  0.145833      0.46  0.000000  1.000000  0.009018   
95410  0.353535  0.202020  0.187500      0.02  0.016667  0.777778  0.028524   
95411  0.222222  0.272727  0.166667      0.20  0.050000  0.888889  0.064286   

         POP902    POP903   POP90C1   POP90C2   POP90C3   POP90C4   POP90C5  \
0      0.017566  0.014033  0.000000  0.353535  0.656566  0.528090  0.697368   
1      0.062546  0.042183  1.000000  0.000000  0.000000  0.561798  0.657895   
2      0.135738  0.112811  0.000000  0.020202  0.989899  0.550562  0.671053   
3      0.010646  0.009257  0.000000  0.080808  0.929293  0.606742  0.605263   
4      0.041719  0.032165  1.000000  0.000000  0.000000  0.516854  0.710526   
...         ...       ...       ...       ...       ...       ...       ...   
95407  0.482534  0.424236  1.000000  0.000000  0.000000  0.561798  0.657895   
95408  0.021425  0.015258  0.969697  0.000000  0.040404  0.573034  0.644737   
95409  0.008716  0.008665  1.000000  0.000000  0.000000  0.595506  0.618421   
95410  0.028744  0.021472  1.000000  0.000000  0.000000  0.528090  0.697368   
95411  0.079313  0.056723  1.000000  0.000000  0.000000  0.539326  0.684211   

           ETH1      ETH2      ETH3      ETH4      ETH5      ETH6      ETH7  \
0      0.929293  0.010101  0.000000  0.000000  0.112245  0.000000  0.000000   
1      0.676768  0.000000  0.000000  0.356322  0.061224  0.363636  0.036364   
2      0.969697  0.020202  0.000000  0.000000  0.020408  0.000000  0.000000   
3      0.616162  0.000000  0.000000  0.126437  0.326531  0.545455  0.036364   
4      0.020202  0.989899  0.000000  0.000000  0.010204  0.000000  0.000000   
...         ...       ...       ...       ...       ...       ...       ...   
95407  0.787879  0.101010  0.076923  0.045977  0.051020  0.000000  0.000000   
95408  0.919192  0.030303  0.000000  0.022989  0.061224  0.090909  0.000000   
95409  0.828283  0.141414  0.000000  0.011494  0.091837  0.000000  0.000000   
95410  0.929293  0.010101  0.012821  0.057471  0.081633  0.000000  0.018182   
95411  0.868687  0.121212  0.000000  0.011494  0.010204  0.000000  0.000000   

           ETH8      ETH9     ETH10  ETH11  ETH12     ETH13     ETH14  \
0      0.000000  0.000000  0.000000    0.0    0.0  0.119565  0.000000   
1      0.157895  0.095238  0.583333    0.0    0.0  0.021739  0.00000

In [25]:
# Concatinating prepared data with predicted Target_B

data_donors = pd.concat([X_df, donors_predicted], axis = 1)
data_donors.head()

TCODE       AGE    INCOME   WEALTH1       HIT  MALEMILI   MALEVET  \
0  0.000000  0.542169  0.666667  1.000000  0.000000  0.000000  0.493671   
1  0.000026  0.373494  0.833333  1.000000  0.066667  0.000000  0.189873   
2  0.000026  0.561586  0.333333  0.111111  0.008333  0.000000  0.253165   
3  0.000000  0.662651  0.000000  0.444444  0.008333  0.000000  0.291139   
4  0.000000  0.759036  0.333333  0.222222  0.250000  0.010417  0.354430   

   VIETVETS  WWIIVETS  LOCALGOV  STATEGOV    FEDGOV   WEALTH2    POP901  \
0  0.343434  0.181818  0.208333      0.04  0.016667  0.555556  0.016206   
1  0.555556  0.111111  0.125000      0.04  0.016667  1.000000  0.058993   
2  0.292929  0.333333  0.125000      0.16  0.016667  0.111111  0.114375   
3  0.141414  0.313131  0.062500      0.00  0.050000  0.000000  0.010456   
4  0.090909  0.535354  0.541667      0.06  0.033333  1.000000  0.041169   

     POP902    POP903  POP90C1   POP90C2   POP90C3   POP90C4   POP90C5  \
0  0.017566  0.014033      0.0  0.353535  0.656566  0.528090  0.697368   
1  0.062546  0.042183      1.0  0.000000  0.000000  0.561798  0.657895   
2  0.135738  0.112811      0.0  0.020202  0.989899  0.550562  0.671053   
3  0.010646  0.009257      0.0  0.080808  0.929293  0.606742  0.605263   
4  0.041719  0.032165      1.0  0.000000  0.000000  0.516854  0.710526   

       ETH1      ETH2  ETH3      ETH4      ETH5      ETH6      ETH7      ETH8  \
0  0.929293  0.010101   0.0  0.000000  0.112245  0.000000  0.000000  0.000000   
1  0.676768  0.000000   0.0  0.356322  0.061224  0.363636  0.036364  0.157895   
2  0.969697  0.020202   0.0  0.000000  0.020408  0.000000  0.000000  0.000000   
3  0.616162  0.000000   0.0  0.126437  0.326531  0.545455  0.036364  0.000000   
4  0.020202  0.989899   0.0  0.000000  0.010204  0.000000  0.000000  0.000000   

       ETH9     ETH10  ETH11  ETH12     ETH13  ETH14     ETH15     ETH16  \
0  0.000000  0.000000    0.0    0.0  0.119565    0.0  0.000000  0.000000   
1  0.095238  0.583333    0.0    0.0  0.021739    0.0  0.015385  0.070175   
2  0.000000  0.000000    0.0    0.0  0.021739    0.0  0.000000  0.000000   
3  0.000000  0.000000    0.0    0.0  0.336957    0.0  0.000000  0.017544   
4  0.000000  0.000000    0.0    0.0  0.000000    0.0  0.000000  0.000000   

     AGE901    AGE902  AGE903    AGE904    AGE905    AGE906    AGE907  \
0  0.493671  0.607595  0.6375  0.526316  0.657895  0.710526  0.438596   
1  0.430380  0.518987  0.5375  0.421053  0.552632  0.592105  0.561404   
2  0.443038  0.544304  0.5750  0.486842  0.592105  0.644737  0.403509   
3  0.405063  0.506329  0.5500  0.447368  0.565789  0.618421  0.438596   
4  0.417722  0.569620  0.6250  0.473684  0.605263  0.657895  0.473684   

      CHIL1     CHIL2     CHIL3     AGEC1     AGEC2     AGEC3    AGEC4  \
0  0.313131  0.626866  0.272727  0.120879  0.218750  0.375000  0.53125   
1  0.333333  0.686567  0.212121  0.142857  0.218750  0.687500  0.71875   
2  0.353535  0.597015  0.252525  0.142857  0.312500  0.395833  0.50000   
3  0.454545  0.522388  0.202020  0.164835  0.390625  0.354167  0.53125   
4  0.343434  0.641791  0.232323  0.153846  0.328125  0.270833  0.46875   

      AGEC5     AGEC6     AGEC7    CHILC1    CHILC2    CHILC3  CHILC4  \
0  0.333333  0.211538  0.227273  0.121212  0.333333  0.507463    0.50   
1  0.256410  0.076923  0.030303  0.111111  0.484848  0.537313    0.44   
2  0.333333  0.192308  0.121212  0.151515  0.424242  0.447761    0.44   
3  0.307692  0.134615  0.106061  0.202020  0.515152  0.447761    0.28   
4  0.512821  0.230769  0.075758  0.131313  0.454545  0.507463    0.38   

     CHILC5    HHAGE1    HHAGE2    HHAGE3      HHN1      HHN2      HHN3  \
0  0.219512  0.265306  0.121951  0.237113  0.191489  0.388235  0.494949   
1  0.182927  0.122449  0.012195  0.051546  0.042553  0.247059  0.757576   
2  0.231707  0.255102  0.121951  0.237113  0.223404  0.411765  0.444444   
3  0.231707  0.255102  0.134146  0.237113  0.244681  0.317647  0.505051   
4  0.231707  0.31

In [26]:
# Now in order to make predictions, how much money will donors donate, I need to choose from data_donors only those,
# where predicted Target_B = 1

In [27]:
data_donors_positive = data_donors[data_donors['TARGET_B_predicted']==1]
data_donors_positive.shape

(21314, 355)

In [28]:
data_donors_positive.head()

TCODE       AGE    INCOME   WEALTH1       HIT  MALEMILI   MALEVET  \
0   0.000000  0.542169  0.666667  1.000000  0.000000  0.000000  0.493671   
2   0.000026  0.561586  0.333333  0.111111  0.008333  0.000000  0.253165   
3   0.000000  0.662651  0.000000  0.444444  0.008333  0.000000  0.291139   
7   0.000000  0.561586  0.166667  1.000000  0.004167  0.000000  0.430380   
12  0.000000  0.686747  0.500000  0.222222  0.066667  0.020833  0.379747   

    VIETVETS  WWIIVETS  LOCALGOV  STATEGOV    FEDGOV   WEALTH2    POP901  \
0   0.343434  0.181818  0.208333      0.04  0.016667  0.555556  0.016206   
2   0.292929  0.333333  0.125000      0.16  0.016667  0.111111  0.114375   
3   0.141414  0.313131  0.062500      0.00  0.050000  0.000000  0.010456   
7   0.363636  0.121212  0.062500      0.06  0.016667  1.000000  0.024260   
12  0.313131  0.333333  0.187500      0.10  0.100000  0.000000  0.023443   

      POP902    POP903  POP90C1   POP90C2   POP90C3   POP90C4   POP90C5  \
0   0.017566  0.014033      0.0  0.353535  0.656566  0.528090  0.697368   
2   0.135738  0.112811      0.0  0.020202  0.989899  0.550562  0.671053   
3   0.010646  0.009257      0.0  0.080808  0.929293  0.606742  0.605263   
7   0.028212  0.021641      1.0  0.000000  0.000000  0.573034  0.644737   
12  0.025551  0.020415      0.0  0.000000  1.000000  0.573034  0.644737   

        ETH1      ETH2  ETH3      ETH4      ETH5      ETH6      ETH7  ETH8  \
0   0.929293  0.010101   0.0  0.000000  0.112245  0.000000  0.000000   0.0   
2   0.969697  0.020202   0.0  0.000000  0.020408  0.000000  0.000000   0.0   
3   0.616162  0.000000   0.0  0.126437  0.326531  0.545455  0.036364   0.0   
7   0.969697  0.020202   0.0  0.011494  0.020408  0.000000  0.000000   0.0   
12  1.000000  0.000000   0.0  0.000000  0.000000  0.000000  0.000000   0.0   

    ETH9  ETH10  ETH11  ETH12     ETH13  ETH14  ETH15     ETH16    AGE901  \
0    0.0    0.0    0.0    0.0  0.119565    0.0    0.0  0.000000  0.493671   
2    0.0    0.0    0.0    0.0  0.021739    0.0    0.0  0.000000  0.443038   
3    0.0    0.0    0.0    0.0  0.336957    0.0    0.0  0.017544  0.405063   
7    0.0    0.0    0.0    0.0  0.010870    0.0    0.0  0.017544  0.354430   
12   0.0    0.0    0.0    0.0  0.000000    0.0    0.0  0.000000  0.392405   

      AGE902  AGE903    AGE904    AGE905    AGE906    AGE907     CHIL1  \
0   0.607595  0.6375  0.526316  0.657895  0.710526  0.438596  0.313131   
2   0.544304  0.5750  0.486842  0.592105  0.644737  0.403509  0.353535   
3   0.506329  0.5500  0.447368  0.565789  0.618421  0.438596  0.454545   
7   0.468354  0.4875  0.394737  0.513158  0.565789  0.561404  0.363636   
12  0.531646  0.5625  0.447368  0.605263  0.644737  0.561404  0.353535   

       CHIL2     CHIL3     AGEC1     AGEC2     AGEC3    AGEC4     AGEC5  \
0   0.626866  0.272727  0.120879  0.218750  0.375000  0.53125  0.333333   
2   0.597015  0.252525  0.142857  0.312500  0.395833  0.50000  0.333333   
3   0.522388  0.202020  0.164835  0.390625  0.354167  0.53125  0.307692   
7   0.626866  0.232323  0.153846  0.437500  0.541667  0.50000  0.256410   
12  0.671642  0.191919  0.131868  0.343750  0.416667  0.40625  0.333333   

       AGEC6     AGEC7    CHILC1    CHILC2    CHILC3  CHILC4    CHILC5  \
0   0.211538  0.227273  0.121212  0.333333  0.507463    0.50  0.219512   
2   0.192308  0.121212  0.151515  0.424242  0.447761    0.44  0.231707   
3   0.134615  0.106061  0.202020  0.515152  0.447761    0.28  0.231707   
7   0.076923  0.030303  0.151515  0.424242  0.537313    0.42  0.170732   
12  0.211538  0.136364  0.131313  0.484848  0.507463    0.44  0.182927   

      HHAGE1    HHAGE2    HHAGE3      HHN1      HHN2      HHN3      HHN4  \
0   0.265306  0.121951  0.237113  0.191489  0.388235  0.494949  0.301075   
2   0.255102  0.121951  0.237113  0.223404  0.411765  0.444444  0.236559   
3   0.255102  0.134146  0.237113  0.244681  0.317647  0.505051  0.322581   
7   0.102041  0.036585  0.092784  0.159574  0.341176  0.565657  0

In [29]:
# Now I need to run my linear model to predict TARGET_D on it.

In [30]:
# lm is the model
# data_donors_positive is the dataset on which it should be performed. 

In [31]:
data_donors_positive
X_new = data_donors_positive.drop(columns=['TARGET_B_predicted'])
X_new

TCODE       AGE    INCOME   WEALTH1       HIT  MALEMILI   MALEVET  \
0      0.000000  0.542169  0.666667  1.000000  0.000000  0.000000  0.493671   
2      0.000026  0.561586  0.333333  0.111111  0.008333  0.000000  0.253165   
3      0.000000  0.662651  0.000000  0.444444  0.008333  0.000000  0.291139   
7      0.000000  0.561586  0.166667  1.000000  0.004167  0.000000  0.430380   
12     0.000000  0.686747  0.500000  0.222222  0.066667  0.020833  0.379747   
...         ...       ...       ...       ...       ...       ...       ...   
95375  0.000026  0.590361  0.166667  0.444444  0.004167  0.000000  0.392405   
95380  0.000000  0.397590  0.666667  1.000000  0.000000  0.000000  0.392405   
95387  0.000718  0.662651  0.666667  0.333333  0.058333  0.000000  0.455696   
95406  0.000000  0.421687  0.833333  0.666667  0.041667  0.020833  0.556962   
95410  0.000000  0.518072  1.000000  1.000000  0.000000  0.000000  0.354430   

       VIETVETS  WWIIVETS  LOCALGOV  STATEGOV    FEDGOV   WEALTH2    POP901  \
0      0.343434  0.181818  0.208333      0.04  0.016667  0.555556  0.016206   
2      0.292929  0.333333  0.125000      0.16  0.016667  0.111111  0.114375   
3      0.141414  0.313131  0.062500      0.00  0.050000  0.000000  0.010456   
7      0.363636  0.121212  0.062500      0.06  0.016667  1.000000  0.024260   
12     0.313131  0.333333  0.187500      0.10  0.100000  0.000000  0.023443   
...         ...       ...       ...       ...       ...       ...       ...   
95375  0.282828  0.202020  0.062500      0.04  0.016667  0.444444  0.091993   
95380  0.202020  0.525253  0.416667      0.26  0.000000  1.000000  0.008021   
95387  0.181818  0.606061  0.145833      0.08  0.016667  1.000000  0.014442   
95406  0.262626  0.464646  0.520833      0.18  0.016667  0.666667  0.029962   
95410  0.353535  0.202020  0.187500      0.02  0.016667  0.777778  0.028524   

         POP902    POP903  POP90C1   POP90C2   POP90C3   POP90C4   POP90C5  \
0      0.017566  0.014033      0.0  0.353535  0.656566  0.528090  0.697368   
2      0.135738  0.112811      0.0  0.020202  0.989899  0.550562  0.671053   
3      0.010646  0.009257      0.0  0.080808  0.929293  0.606742  0.605263   
7      0.028212  0.021641      1.0  0.000000  0.000000  0.573034  0.644737   
12     0.025551  0.020415      0.0  0.000000  1.000000  0.573034  0.644737   
...         ...       ...      ...       ...       ...       ...       ...   
95375  0.103666  0.087409      0.0  0.000000  1.000000  0.573034  0.644737   
95380  0.009249  0.008158      1.0  0.000000  0.000000  0.584270  0.631579   
95387  0.017100  0.016696      1.0  0.000000  0.000000  0.471910  0.763158   
95406  0.035199  0.026544      0.0  0.000000  1.000000  0.550562  0.671053   
95410  0.028744  0.021472      1.0  0.000000  0.000000  0.528090  0.697368   

           ETH1      ETH2      ETH3      ETH4      ETH5      ETH6      ETH7  \
0      0.929293  0.010101  0.000000  0.000000  0.112245  0.000000  0.000000   
2      0.969697  0.020202  0.000000  0.000000  0.020408  0.000000  0.000000   
3      0.616162  0.000000  0.000000  0.126437  0.326531  0.545455  0.036364   
7      0.969697  0.020202  0.000000  0.011494  0.020408  0.000000  0.000000   
12     1.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
...         ...       ...       ...       ...       ...       ...       ...   
95375  0.969697  0.000000  0.038462  0.000000  0.020408  0.000000  0.000000   
95380  0.959596  0.010101  0.000000  0.022989  0.020408  0.000000  0.000000   
95387  0.979798  0.020202  0.000000  0.000000  0.000000  0.000000  0.000000   
95406  0.888889  0.080808  0.012821  0.011494  0.030612  0.000000  0.000000   
95410  0.929293  0.010101  0.012821  0.057471  0.081633  0.000000  0.018182   

           ETH8  ETH9     ETH10  ETH11  ETH12     ETH13  ETH14  ETH15  \
0      0.000000   0.0  0.000000    0.0    0.0  0.119565    0.0    0.0   
2      0.000000   0.0  0.000000    0.0    0.0  0.021739    0.0    0.0   
3      

In [32]:
predictions_for_amount = lm.predict(X_new)

In [37]:
# predictions_for_amount

In [34]:
predictions_for_amount = pd.DataFrame(predictions_for_amount)

In [35]:
predictions_for_amount.head(10)

0
0   7.327936
1   5.681951
2   5.316482
3  11.657956
4   2.639375
5   5.369431
6  13.878977
7  12.923051
8   8.854602
9  11.895283

In [38]:
predictions_for_amount = predictions_for_amount.rename(columns={0:'TARGET_D_predicted'})
predictions_for_amount.head()

TARGET_D_predicted
0            7.327936
1            5.681951
2            5.316482
3           11.657956
4            2.639375

In [ ]:
# Calculation how much money will this email campaign bring
# sum(all_donations) - (all sent emails)*0.68 

In [41]:
win_sent_aimed = predictions_for_amount["TARGET_D_predicted"].sum() - (predictions_for_amount["TARGET_D_predicted"].count()*0.68)

In [42]:
win_sent_aimed

220619.39173452597

In [43]:
# if we sent emails to those, who will not donate at all:
win_sent_to_all = predictions_for_amount["TARGET_D_predicted"].sum() - (data_donors["TARGET_B_predicted"].count()*0.68)
win_sent_to_all

170232.75173452595

In [44]:
220619.39173452597-170232.75173452595

50386.640000000014

#### Final conclusions:
- Sending emails to all people will cost more money. For each sent email it will cost 0.68 dolars. .
- With sending emails only to those people, who are more likely to be donors we can save sending costs. 
- If we assume, that we sent emails only to people, who are possible donors, then it means we can save costs and have from each send email a win of money. 
- In the business context it is better to take into consideration sending emails to people more likely to be donors.
- In total this email campaign brought 220619.39 dolars win.
- If we have sent emails to all from the list, then we will win 170232.75 dolars.
- Delta is 50386.64 dolars so this is what the company will loose, if emails were sent to all (naive scenario from Monday). 